### PR TITLE
For #19933 - in progress media tab

### PR DESCRIPTION
**Pull request by Mugurell**
Since the current ui/ux assumes the same as with the existing recent tab I reused most of the existing work.
When later requirements arrive the current implementation can be updated.

Needs https://github.com/mozilla-mobile/android-components/pull/10518

Desired UI - https://www.figma.com/file/am9EJGg6ow29PXWkc9jcOH/MR2.1?node-id=220%3A125
More details in https://jira.mozilla.com/browse/FNX-22337

https://user-images.githubusercontent.com/11428869/123232731-1c741880-d4e2-11eb-9030-7608e1e111fc.mp4

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by Mugurell** [comment]
Request for data collection review
---
1. What questions will you answer with this data?
    Usefulness and user engagement with a new feature.
 
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
    Data will be used to improve this new feature.
 
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
    No alternative is available as this is a new functionality of the app.
 
4. Can current instrumentation answer these questions?
    No.
 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement   Description | Data Collection   Category | Tracking Bug #
-- | -- | --
Recent tabs - "Show all" click to see all opened normal tabs | category 2 - Interaction data | https://github.com/mozilla-mobile/fenix/issues/19993
Recent tabs - normal tab click to go back to it in browser| category 2 - Interaction data | https://github.com/mozilla-mobile/fenix/issues/19993
Recent tabs - media tab click to go back to the last tab playing media| category 2 - Interaction data | https://github.com/mozilla-mobile/fenix/issues/19993

6. How long will this data be collected? Choose one of the following:
    Data will be collected for 6 months initially (ending on 2021-12-01), with the option to renew at that point. 
 
7. What populations will you measure?
    No filters - all locales, channels, etc once the feature rides the trains.
 
8. If this data collection is default on, what is the opt-out mechanism for users?
     Default Glean SDK opt-out mechanism.
 
9. Please provide a general description of how you will analyze this data.
    Glean / Amplitude
 
10. Where do you intend to share the results of your analysis?
      Only on Glean, Amplitude and with mobile teams.
 
11. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
      No third-party tool will use this data.

**Comment by Mugurell** [comment]
The UI/UX should be ready for review.
But let's wait with the code review for after the PR on AC is approved.

**Comment by gabrielluong** [comment]
We'll probably got to this review from both Eng and UX after the break.

**Review by jonalmeida — COMMENTED** [review]
Some short comments that would not be affected by any changes by the AC PR review.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt:106]: There is a different background for each item depending on the position in the "Jump back in section".
In all other cases we use a drawable with rounded corners.
For a middle item we don't need rounded corners, just the right color.

[top, middle, middle, bottom] items would look like:

image

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:239]: I envisioned a much greater importance for this item, really a separate section, much more powerful and customizable than just tabs listing but will demote it to that status.

Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:239]: I think that's the right mentality for a previous design iteration where we had a specific section for "jumping back in" to a media tab.

For now at least, it looks like we want to group them under one section so we can simplify the work for now and come back to it if that changes.

File: app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt:33]: Will add some tests for these methods also.

File: app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt:39]: nit: a property extension might be more consistent with the other ones used in AC already, like `selectedNormalTab`, `normalTabs`, etc.

```suggestion
val BrowserState.inProgressMediaTab: TabSessionState?
    get() = normalTabs
        .filter { it.lastMediaAccess > 0 }
        .maxByOrNull { it.lastMediaAccess 
```

File: app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt:22]: nit: I mentioned this to Elise as well, I wonder if we all have different IDE settings that think the max line is shorter than what we have set on detekt. 🤔 I think this IDE plugin might help: https://github.com/detekt/detekt-intellij-plugin

This should be able to fit on one line. 

```suggestion
    val selectedNormalTab = selectedNormalTab ?: normalTabs.maxByOrNull { it.lastAccess }
```

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt:33]: We don't want to remove the observers of the title/icon to prevent a regression of this bug: https://github.com/mozilla-mobile/fenix/issues/20054

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt:49]: I like this solution a lot for handling the various layouts. 🙂

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt:106]: Doesn't the colour stay the same rounded or not? I would have expected the background drawable to have only changed here. If I'm misunderstanding, please ignore.

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlView.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlView.kt:74]: nit: VisibleForTesting should go right after here instead.

**Review by jonalmeida — COMMENTED** [review]
Looks good with the update! There is just one concern in the feature that I think we should address. 

This [AC bump](https://github.com/mozilla-mobile/fenix/pull/20312) should have your changes as well, but we can see to that on Monday. 🙂

File: app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/ext/BrowserState.kt:22]: I intended to write this on two lines. Will make it a one liner instead.

This line will be in the getter of a new `val BrowserState.lastOpenedNormalTab` since we need to observe id/title/icon of it.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/RecentTabsListFeature.kt:33]: Thank you!
Will then observe `selectedNormalTab?.id`, `selectedNormalTab?.content?.title`, `selectedNormalTab?.content?.icon` and `normalTabs`.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt:106]: I went with not setting any background for the items in xml and then setting a different one for each variation in code.
If not setting a background here this would be the results:
| dark theme | light theme |
| ----------- | ----------- |
| image | image |

I find it cleaner and more helpful having all customization used in one file making it easy to spot the differences.
Can also set that color as `background` in xml and then do nothing here but overwrite it in all other cases.
Or are you proposing a yet different approach?

Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt:106]: Having `android:src="?attr/above"` set in xml and then use 
> AppCompatResources.getDrawable(context, R.drawable.rounded_bottom_corners)

in code would work but the current `R.drawable.rounded_bottom_corners` (& co) are shapes also declaring a background color so to keep the implementation clean we'd need to create new shapes just for the rounded corners while allowing the colors to be set separately.

Do you propose going on this route?

**Review by topotropic — COMMENTED** [review]
Looks good to me based on the video, I wasn't able to install the apk from Taskcluster.

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabViewDecorator.kt:106]: > I find it cleaner and more helpful having all customization used in one file making it easy to spot the differences.

I agree! I hope we can get this right when we migrate to Compose.

> Do you propose going on this route?

No, I think this is fine. Thank you for your detailed explanation. 🙂

**Comment by jonalmeida** [comment]
Latest [AC bump](https://github.com/mozilla-mobile/fenix/pull/20356) went through, I think we can rebase this now.

**Review by jonalmeida — APPROVED** [review]
✨

**Pull request merged**
